File: Assets/Player.js
#pragma strict

public var gun : GameObject;
public var damage : float = 0.05;
public var mask : LayerMask;
private var fire : boolean = false;
private var stats : PlayerStats = new PlayerStats();

function Start(){
	stats.level = PlayerPrefs.GetInt("level");
	stats.xp = PlayerPrefs.GetInt("xp");
	stats.strength = PlayerPrefs.GetInt("strength");
}

function FixedUpdate () {
	if(Input.GetKeyDown(KeyCode.Escape)) {
		Application.Quit();
	}
			
	// flamethrower toggle
	if(Input.GetButtonDown("Fire1")) fire = !fire;
	if(Input.GetButton("Fire2")) fire = false;
		
	if(fire || Input.GetButton("Fire1")) {

		gun.particleSystem.Play();
		gun.audio.volume = gun.audio.volume * 0.95 + 0.05 * 0.1; // max volume
		
		
		var hit : RaycastHit;
		var jitter : Vector3 = gun.transform.right * (Random.value - 0.5) + gun.transform.up * (Random.value - 0.5);
		var direction : Vector3 = gun.transform.forward + jitter * 0.6;
		var distance : float = 6;
        if (Physics.Raycast(gun.transform.position, direction, hit, distance, mask)) {
        	Debug.DrawRay(gun.transform.position, direction * distance, Color.red, 0.1); 
        	hit.collider.transform.localScale.y *= (1 - damage);
	   	    hit.collider.renderer.material.color = Color.black;
            if(hit.collider.transform.localScale.y < 0.25) {
            	stats.addXP(2);
	        	Destroy(hit.transform.gameObject);
	        	Debug.Log(stats.getXP());
            }
        }
        
	} else {
		gun.particleSystem.Stop();
		gun.audio.volume = gun.audio.volume * 0.9 + 0.1 * 0; // min volume
	}

	if(gameObject.transform.localScale.y < 1) {
		var health : float = gameObject.transform.localScale.y * 0.99 + 0.01;
		var heal : float = health - gameObject.transform.localScale.y;
		gameObject.transform.position.y += heal;
		gameObject.transform.localScale.y = health;
	}
}

File: Assets/Player.cs
using UnityEngine;
using System.Collections;

public class Player : Photon.MonoBehaviour {

	public float damage = 0.05f;
	public LayerMask mask;
	private bool want_fire = false;
	private bool firing = false;
	private bool bleeding = false;

	private Vector3 correctPos = Vector3.zero;
	private Quaternion correctRot = Quaternion.identity;
	
	void Start()
	{
		if(photonView.isMine) {
			// TODO enable collision event emission on particleSystem only if photonView.isMine
		}
	}


    void Update()
    {
        if (!photonView.isMine)
		{
            transform.position = Vector3.Lerp(transform.position, this.correctPos, Time.deltaTime * 5);
            transform.rotation = Quaternion.Lerp(transform.rotation, this.correctRot, Time.deltaTime * 5);
        }
    }
	
	void OnPhotonSerializeView(PhotonStream stream, PhotonMessageInfo info)
    {
        if (stream.isWriting)
        {
            stream.SendNext(transform.position);
            stream.SendNext(transform.rotation);
        }
        else
        {
            this.correctPos = (Vector3)stream.ReceiveNext();
            this.correctRot = (Quaternion)stream.ReceiveNext();
        }
    }
	
	

	
	void FixedUpdate () {
	
		if(photonView.isMine) {
			// flamethrower toggle
			if(Input.GetButtonDown("Fire1")) want_fire = !want_fire;
			if(Input.GetButton("Fire2")) want_fire = false;
				
			if(firing != want_fire) {
				firing = want_fire;
				if(want_fire) {
					photonView.RPC("FireStart", PhotonTargets.All);
				} else {
					photonView.RPC("FireStop", PhotonTargets.All);
				}
			}
		}

		// healing
		if(!bleeding && transform.localScale.y < 1) {
			renderer.material.color = Color.white;

			float health = transform.localScale.y * 0.99f + 0.01f;
			float heal = health - transform.localScale.y;
			
			transform.position += Vector3.up * heal;
			
			Vector3 scale = transform.localScale;
			scale.y = health;
			transform.localScale = scale;
		}
		bleeding = false;
	}

	void OnParticleCollision (GameObject other) {
		if(photonView.isMine && other.layer == 8) { // me hitting enemy
			PhotonView pv = other.GetComponent("PhotonView") as PhotonView;
			if(pv) {
				pv.RPC("Burn", PhotonTargets.All, damage);
			}
		}
	}


	[RPC]
	void FireStart() {
		particleSystem.Play();
		audio.volume = audio.volume * 0.95f + 0.05f * 0.1f; // max volume
	}
	
	[RPC]
	void FireStop() {
		particleSystem.Stop();
		audio.volume = audio.volume * 0.9f + 0.1f * 0f; // min volume
	}
	
	[RPC]
	void Bitten (float damage) {
		bleeding = true;
		renderer.material.color = Color.black;
		if(photonView.isMine) {
			Vector3 scale = transform.localScale;
			scale.y = scale.y * (0.98f + 0.02f * (1.0f - damage)); // the smaller the enemy, the less the damage  // FIXME damage hardcoded
			if(scale.y < 0.25f) {
				scale.y = 1.0f;
				transform.position = Vector3.up * 7.0f;
			}
			transform.localScale = scale;
		}
	}

}
